lib/charms/layer/openstack.py

```diff
@@ -305,6 +303,16 @@ def _default_subnet(members):
         raise OpenStackLBError(action='create', exc=False)
 
 
+def _is_base64(s):
+    """
+    Verify is the utf8 encoded string is base64 encoded or not
+    """
+    try:
+        return b64encode(b64decode(s)) == s
+    except Exception:
+        return False
```

Comment:
I'm not sure that this is entirely safe, since there are different variants of base64 and variations between encodings which could still be valid, so the decode-encode might not be 1-to-1 exact.  But I'm also not at all sure if the false negatives would be more common than the current false positive, so this is probably better than what we have and worth going forward with.  I certainly don't have a better suggestion for trying to detect the encoding.  I wouldn't mind another perspective on it, though.

Reply:
Agreed. A base64 string with a newline in it will fail the test:

```
>>> s = "Zm9v\nCg=="
>>> b64decode(s)
b'foo\n'
>>> b64encode(b64decode(s))
b'Zm9vCg=='
>>> b64encode(b64decode(s)) == s
False
```

Do we have a guarantee that the `ca_cert` string, when it's coming in as base64, won't include newlines?

Reply:
No. That's quite the opposite. The `ca_cert` b64encoded will contain new lines.

Reply:
@Cynerva But that does not cause issues as the `ca_cert` is first encoded in `utf8`. See the additional assert that I added in my tests.

Reply:
Ah, right. For that last comparison to make sense, I should have utf8-encoded the string so both sides are bytes, not strings. Sorry about that.

But that doesn't resolve my concern. Try with `b'Zm9vYm\nFyCg=='` instead and you'll see what I mean. It is valid base64:

```
>>> b64decode(b'Zm9vYm\nFyCg==')
b'foobar\n'
```

but does not pass the test:

```
>>> s = b'Zm9vYm\nFyCg=='
>>> b64encode(b64decode(s)) == s
False
```

To be clear, it is the newline in `b'Zm9vYm\nFyCg=='` that is a problem, not the newline in `b'foobar\n'`.

Reply:
@Cynerva, noted. I updated the `_is_base64()` function to take new lines into account. Thanks for the feedback!

Reply:
Thanks. LGTM